[jordon0802] iP

jordon0802 opened the pull request:
# Dominic

## Features:
- _Add_ Todos/Deadlines/Events
- _Mark completed_ Todos/Deadlines/Events
- _Delete_ Todos/Deadlines/Events

## Why should you use it?
1. It's user friendly😄 
2. No personal data collected
3. It's **free**!!

### Some users have commented:
> "this is the best chat bot assistant I have used!!"

### Take a sneak peek at the ```main``` method
```java
public static void main(String[] args) {
    // Initialize Storage File
    boolean success = StorageReader.isInitialized();
    if (!success) {
        return;
    }

    // Greet message
    Dominic.greet();
    
    // ...
}
```

### Features implemented:
- [x] [Gradle](https://gradle.org/) support
- [ ] Checkstyle
- [ ] Java FX

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jordon0802 committed:
Rename Duke.java to Dominic.java

jordon0802 committed:
Dominic.java: Add echo feature and edit greet message

jordon0802 committed:
Add List.java

jordon0802 committed:
Dominic.java: Add list function

jordon0802 committed:
Add Task.java

jordon0802 committed:
List.java: Modify List to become a list of tasks

jordon0802 committed:
Add mark and unmark feature

jordon0802 committed:
Add ToDos.java

jordon0802 committed:
Add Deadlines.java

jordon0802 committed:
Add Events.java

jordon0802 committed:
Task.java: Change some access modifiers to accomodate new classes

jordon0802 committed:
Dominic.java: Add support for ToDos, Events and Deadlines

jordon0802 committed:
Dominic.java: Remove trailing whitespaces and edit typo

jordon0802 committed:
Rename Duke to Dominic

jordon0802 committed:
input.txt: Add inputs for testing

jordon0802 committed:
EXPECTED.TXT: Add expected outputs for testing

jordon0802 committed:
Add custom exceptions

jordon0802 committed:
Dominic.java: Add clearer exception messages

jordon0802 committed:
Add InvalidKeywordOrderException.java

jordon0802 committed:
Add delete task feature

jordon0802 committed:
Refactor so that events and deadlines store dates

Subclasses of Task receives entire input of user.

Tasks should contain only relevant information regarding the event
(i.e. command word and keywords are redundant).

Let's breakdown input into sections, removing command words and
keywords, before passing them into the constructors. Subclasses of
Task will now store the relevant dates while the Task class store
the task.

Leaving the job to the subclasses of Task, causes unexpected errors
in the toString() method.

jordon0802 committed:
Dominic.java: Use enum and switch instead of if-else

jordon0802 committed:
Add InvalidFileFormatException.java

jordon0802 committed:
Task.java: Make Task.java abstract

jordon0802 committed:
Task subclasses: Implement toFileString() and getValid[Task]() methods

jordon0802 committed:
Dominic.java: Implement save feature

jordon0802 committed:
Merge branch 'branch-level-7'

jordon0802 committed:
ToDos.java: Rename getValidToDo() to getValidTask()

jordon0802 committed:
Task.java: Add support for LocalDate to String conversion

jordon0802 committed:
Add support for Date

jordon0802 committed:
Add checks for start and end date

jordon0802 committed:
Add filter by date feature

jordon0802 committed:
Merge branch 'branch-level-8'

jordon0802 committed:
Dominic.java Code logic is mainly in Dominic.java

OOP principles are not adhered to

Let's pull up storage related methods to the StorageReader and
StorageWriter class and package it

Let's pull up date related methods to the utility class DateFormatter

Let's pull up parsing related code to the utility class Parser

Let's create the Command class which will be the parent class of all
known commands

Let's package all the subpackages under the main package, dominic

Adhering to single responsibility principle is essential

jordon0802 committed:
StorageReader.java: Use enum to read task type

jordon0802 committed:
Merge branch 'add-gradle-support'

0x4F776C commented on src/main/java/dominic/commands/EventCommand.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.time.LocalDate;
+
+public class EventCommand extends Command {
```
LGTM! Well-structured and naming convention matches requirements

0x4F776C commented on src/main/java/dominic/commands/FilterCommand.java:
```diff
@@ -0,0 +1,44 @@
+    public void execute() {
+        if (DateFormatter.isLocalDate(super.getArguments())) {
+            LocalDate date = DateFormatter.toLocalDate(super.getArguments());
+            Task[] list = List.toTaskArray();
```
Can we rename this match the use case (e.g. tasks)

0x4F776C commented on src/main/java/dominic/commands/FilterCommand.java:
```diff
@@ -0,0 +1,44 @@
+            int length = list.length;
+            int counter = 1;
+            for (int i = 1; i <= length; i++) {
+                if (list[i - 1] instanceof Deadline d) {
```
Might want to rename "d" to match the use case

0x4F776C commented on src/main/java/dominic/commands/MarkCommand.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public void execute() {
+        try {
+            int x = Integer.parseInt(super.getArguments());
```
Might want to rename "x" to meet the use case

0x4F776C commented on src/main/java/dominic/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,26 @@
+    public void execute() {
+        try {
+            int y = Integer.parseInt(super.getArguments());
+            Task[] arr1 = List.toTaskArray();
```
Might want to rename "arr1". There aren't other arrays in this code space

0x4F776C commented on src/main/java/dominic/storage/StorageReader.java:
```diff
@@ -0,0 +1,149 @@
+        }
+    }
+
+    public static boolean init() {
```
Not a naming quality violation. But the variable should reflect like a boolean

0x4F776C commented on src/main/java/dominic/utils/Parser.java:
```diff
@@ -0,0 +1,75 @@
+    private Parser() {
+    }
+
+    public static boolean parseInput(String input) {
```
Might want to rename the boolean method

0x4F776C reviewed (commented):
LGTM! However, please take note of the boolean naming conventions

jordon0802 committed:
Task.java: Make checkKeyword() use regex

jordon0802 committed:
Implement JUnit Testing

jordon0802 committed:
Correct syntactical errors

jordon0802 committed:
Add javadoc comments

jordon0802 committed:
Rearrange to adhere to coding standards

jordon0802 committed:
FindCommand.java: Add FindCommand.java

jordon0802 committed:
FindCommand.java: Use regex to find word

jordon0802 committed:
Merge branch 'branch-A-JavaDoc'

jordon0802 committed:
Merge branch 'branch-A-CodingStandard'

jordon0802 committed:
Merge branch 'branch-Level-9'

sherwinee commented on src/main/java/dominic/commands/ByeCommand.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class ByeCommand extends Command {
+    /** Command keyword. */
+    public static final String COMMAND = "bye";
```
I like your use of SCREAMING_SNAKE_CASE for constants in your derived classes of the Command class

sherwinee commented on src/main/java/dominic/storage/StorageReader.java:
```diff
@@ -0,0 +1,164 @@
+package dominic.storage;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.time.LocalDate;
+
+import dominic.exceptions.InvalidFileFormatException;
+import dominic.exceptions.MissingKeywordException;
+import dominic.exceptions.InvalidKeywordOrderException;
+import dominic.exceptions.MissingArgumentException;
+import dominic.tasks.Deadline;
+import dominic.tasks.Event;
+import dominic.tasks.Task;
+import dominic.tasks.Todo;
+import dominic.utils.DateFormatter;
+import dominic.utils.List;
```
I like your consistent ordering of multiple import statements here

sherwinee commented on src/main/java/dominic/utils/Parser.java:
```diff
@@ -0,0 +1,93 @@
+            case ListCommand.COMMAND:
+                ListCommand listCommand = new ListCommand(arguments);
+                listCommand.execute();
+                break;
+            case FilterCommand.COMMAND:
+                FilterCommand filterCommand = new FilterCommand(arguments);
+                filterCommand.execute();
+                break;
+            case DeleteCommand.COMMAND:
+                DeleteCommand deleteCommand = new DeleteCommand(arguments);
+                deleteCommand.execute();
+                break;
+            case MarkCommand.COMMAND:
+                MarkCommand markCommand = new MarkCommand(arguments);
+                markCommand.execute();
+                break;
+            case UnmarkCommand.COMMAND:
+                UnmarkCommand unmarkCommand = new UnmarkCommand(arguments);
+                unmarkCommand.execute();
+                break;
+            case TodoCommand.COMMAND:
+                TodoCommand todoCommand = new TodoCommand(arguments);
+                todoCommand.execute();
+                break;
+            case DeadlineCommand.COMMAND:
+                DeadlineCommand deadlineCommand = new DeadlineCommand(arguments);
+                deadlineCommand.execute();
+                break;
+            case EventCommand.COMMAND:
+                EventCommand eventCommand = new EventCommand(arguments);
+                eventCommand.execute();
+                break;
+            case FindCommand.COMMAND:
+                FindCommand findCommand = new FindCommand(arguments);
+                findCommand.execute();
+                break;
+            // Invalid Command
+            default:
+                System.out.println("Error: Invalid command.");
+            }
```
I like how your switch statement follows the (slightly unconventional) indentation style required in this course

sherwinee commented on src/main/java/dominic/storage/StorageReader.java:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @return true if storage directory exists or is created successfully, otherwise false
+     */
+    public static boolean isInitialized() {
```
I like how your boolean method is named to sound like booleans (isXxx)

sherwinee reviewed (commented):
LGTM! Honestly I have nothing much to say as you seem to have followed the course's Java coding standard fully, I've pointed out some of the adherences to the standard.

jordon0802 committed:
Task.java: Change access modifier

jordon0802 committed:
Add checkstyle plugin

jordon0802 committed:
Merge branch 'branch-A-CheckStyle'

jordon0802 committed:
Add GUI for Dominic

Refactor return values of command so that can be displayed in chat

jordon0802 committed:
Merge branch 'branch-level-10'

jordon0802 committed:
DeadlineTest.java: Add DeadlineTest.java
TodoTest.java: Add testing for toString() methods

jordon0802 committed:
Add guard clause before decrementing size variable

Decrement of size variable occurs before removing Task

Attempting to remove a task from an empty list would result in size
variable becoming negative

Let's check if the index is zero or more and less than size
Let's add try-catch clauses in ListTest.java

This will prevent disparity between size variable and the actual size
of the list

jordon0802 committed:
Dominic.java: Add assert feature

jordon0802 committed:
List.java: Remove redundant getSize() method

jordon0802 committed:
Merge pull request #2 from jordon0802/branch-A-Assertions

Add assert statement

jordon0802 committed:
Merge branch 'master' into branch-A-CodeQuality

jordon0802 committed:
Task child classes: Extract duplicate methods

Multiple methods make use of common code within their methods

This results in code duplication

Let's pull up code into taskStringToDeadline() and taskStringToEvent()
methods respectively

Creating deadline and event from a task string should be done in
Deadline and Event class respectively

jordon0802 committed:
Merge pull request #3 from jordon0802/branch-A-CodeQuality

Improve code readability

jordon0802 committed:
TaskTest.java: Rearrange imports

jordon0802 committed:
Set up GitHub Actions

jordon0802 committed:
Add JavaDoc comments to new classes and methods

jordon0802 committed:
Add archive extension

Tasks that have passed or completed can only be deleted and can no
longer be retrieved

This results in the loss of old tasks

Let's add an archive command that allows completed tasks to be archived

Old tasks can now be still accessed while the list shows currently
pending tasks

jordon0802 committed:
Merge branch 'branch-C-Archive'

jordon0802 committed:
Add writing archive file to local disk

jordon0802 committed:
Update gradle.yml

jordon0802 committed:
Update checkstyle.xml

jordon0802 committed:
StorageReader.java: Change access modifier of FileChoice

jordon0802 committed:
Add Dominic.png

jordon0802 committed:
Add User.png

jordon0802 committed:
Remove DaUser.png

jordon0802 committed:
Remove DaDuke.png

jordon0802 committed:
Rename main.css to main-window.css

jordon0802 committed:
Split dialog-box.css for user and dominic

The dialog-box.css file previously handled styles for both dominic
and the user

This reduces customizability for each of the dialog box

Let's separate the dialog-box.css file to user-dialog-box.css and
dominic-dialog-box.css

This allows for clearer separation of concerns and simplifies future
style modifications for each dialog box

jordon0802 committed:
Split DialogBox.fxml to separate files

jordon0802 committed:
MainWindow.fxml: Rename referenced stylesheets

jordon0802 committed:
MainWindow.java: Change referenced images

jordon0802 committed:
DialogBox.java: Load respective fxml files for dialog box

jordon0802 committed:
Dominic.java: Loop initialization process of storage file

jordon0802 committed:
Merge branch 'branch-A-BetterGui'

jordon0802 committed:
Main.java: Add title to gui

jordon0802 committed:
Add Ui.png

jordon0802 committed:
MarkCommand.java: Add tab character to message
UnmarkCommand.java: Add tab character to message
MainWindow.java: Add early return statement upon exit

jordon0802 committed:
Update README.md

jordon0802 committed:
Merge branch 'branch-A-UserGuide'

jordon0802 committed:
Update README.md

jordon0802 committed:
bug fix: Close file writers after use

jordon0802 committed:
Dominic.java: Add newline to message

jordon0802 committed:
Update versioning

jordon0802 committed:
Remove semicolon typo

jordon0802 committed:
Change font to Courier monospace

jordon0802 committed:
Change user font to Courier monospace

jordon0802 committed:
Add help command feature

jordon0802 committed:
Add feature to write to disk when gui is closed

jordon0802 committed:
Add archive command to help command list

jordon0802 committed:
FilterCommand.java: Add message if no task matched

jordon0802 committed:
FindCommand.java: Add message if no task matched

jordon0802 committed:
ListCommand.java: Add message if no tasks in list

jordon0802 committed:
HelpCommand.java: Add javadoc comment

jordon0802 committed:
Update versioning to v1.1
